src/main/kotlin/com/curiouscreature/kotlin/math/Matrix.kt

```diff
@@ -386,7 +386,78 @@ fun transpose(m: Mat4) = Mat4(
         Float4(m.x.w, m.y.w, m.z.w, m.w.w)
 )
 fun inverse(m: Mat4): Mat4 {
-    TODO("Implement inverse(Mat4)") // TODO
+    val result = Mat4()
+    val pairs = FloatArray(12)
+
+    pairs[0]  = m.z.z * m.w.w
+    pairs[1]  = m.w.z * m.z.w
+    pairs[2]  = m.y.z * m.w.w
+    pairs[3]  = m.w.z * m.y.w
+    pairs[4]  = m.y.z * m.z.w
+    pairs[5]  = m.z.z * m.y.w
+    pairs[6]  = m.x.z * m.w.w
+    pairs[7]  = m.w.z * m.x.w
+    pairs[8]  = m.x.z * m.z.w
+    pairs[9]  = m.z.z * m.x.w
+    pairs[10] = m.x.z * m.y.w
+    pairs[11] = m.y.z * m.x.w
+
+    result.x.x = pairs[0] * m.y.y + pairs[3] * m.z.y + pairs[4] * m.w.y
+    result.x.x -= pairs[1] * m.y.y + pairs[2] * m.z.y + pairs[5] * m.w.y
```

Comment:
Would you mind aligning all the operations? I usually don't align code vertically but in cases like this it makes bugs etc. easier to spot later on. The code should look like this:

```
result.x.y  = pairs[1] * m.x.y + pairs[6] * m.z.y + pairs[9] * m.w.y
result.x.y -= pairs[0] * m.x.y + pairs[7] * m.z.y + pairs[8] * m.w.y
```